ts/async_service.py

```diff
@@ -0,0 +1,156 @@
+            "model: %s, Invalid return type: %s.",
+            context.model_name,
+            type(ret),
+        )
+        return create_predict_response(
+            None, req_id_map, "Invalid model predict output", 503
+        )
+
+    if len(ret) != len(input_batch):
+        logger.warning(
+            "model: %s, number of batch response mismatched, expect: %d, got: %d.",
+            context.model_name,
+            len(input_batch),
+            len(ret),
+        )
+        return create_predict_response(
+            None, req_id_map, "number of batch response mismatched", 503
+        )
+
+    duration = round((time.time() - start_time) * 1000, 2)
+    metrics.add_time(PREDICTION_METRIC, duration)
+
+    return create_predict_response(
+        ret, req_id_map, "Prediction success", 200, context=context
+    )
+
+
+class AsyncService(object):
+    def __init__(self, service):
+        self.service = service
+        self.service.predict = types.MethodType(predict, self.service)
+        self.in_queue = Queue()
+        self.out_queue = None
+        self.loop = None
+
+    def receive_requests(self):
+        while True:
+            logging.debug("Waiting for new message")
+            cmd, msg = retrieve_msg(self.service.cl_socket)
+
+            if cmd == b"I":
+                logging.debug(f"Putting msg in queue: {msg}")
+                self.in_queue.put(msg)
+            else:
+                logging.debug(f"Unexpected request: {cmd}")
+
+    async def call_predict(self, batch):
+        response = await self.service.predict(batch)
+        await self.out_queue.put(response)
+
+    def fetch_batches(self):
+        MAX_WAIT = 0.1
+        BATCH_SIZE = 1
+        while True:
+            st = time.time()
+            batch = []
+            try:
+                request = self.in_queue.get()
+                batch += request
+                while len(batch) < BATCH_SIZE and (time.time() - st) < MAX_WAIT:
+                    timeout = max(0, MAX_WAIT - (time.time() - st))
+                    request = self.in_queue.get(timeout=timeout)
+                    batch += request
+            except Empty:
+                pass
+
+            asyncio.run_coroutine_threadsafe(self.call_predict(batch), self.loop)
+
+    def send_responses(self):
+        while True:
+            future = asyncio.run_coroutine_threadsafe(self.out_queue.get(), self.loop)
+            self.service.cl_socket.sendall(future.result())
+
+    def run(self):
+        async def main():
+            self.loop = asyncio.get_running_loop()
+            self.out_queue = (
+                AsyncQueue(loop=self.loop)
+                if sys.version_info <= (3, 9)
+                else AsyncQueue()
+            )
+            fetch = Thread(target=self.fetch_batches)
+            fetch.start()
+            receive = Thread(target=self.receive_requests)
+            receive.start()
+            send = Thread(target=self.send_responses)
+            send.start()
```

Comment:
If one or more of these threads exit/terminate and the others continue to run, what would the recovery look like? Would we have to restart the thread or somehow trigger an entire worker restart to ensure simple recovery?

Reply:
Good point, let me see if I can make this more fail resistant.